operator/pkg/service/envoy_listener.go

```diff
@@ -0,0 +1,150 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright 2019-2020 Authors of Cilium
+
+package service
+
+import (
+	"fmt"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+
+	"github.com/cilium/cilium/pkg/k8s"
+	"github.com/cilium/cilium/pkg/k8s/informer"
+	slim_corev1 "github.com/cilium/cilium/pkg/k8s/slim/k8s/api/core/v1"
+	"github.com/cilium/cilium/pkg/k8s/utils"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+)
+
+type serviceAddedEvent struct {
+	service *slim_corev1.Service
+}
+
+type serviceUpdatedEvent struct {
+	oldService *slim_corev1.Service
+	newService *slim_corev1.Service
+}
+
+type serviceDeletedEvent struct {
+	service *slim_corev1.Service
+}
+
+type EnvoyLBController struct {
+	envoyConfigManager *envoyConfigManager
+
+	queue      workqueue.RateLimitingInterface
+	maxRetries int
+
+	serviceInformer cache.Controller
+	serviceStore    cache.Store
+}
+
+// NewEnvoyLBController returns a new EnvoyLBController
+func NewEnvoyLBController() (*EnvoyLBController, error) {
+	manager := &EnvoyLBController{
+		queue:      workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
+		maxRetries: 10,
+	}
+	manager.serviceStore, manager.serviceInformer = informer.NewInformer(
+		utils.ListerWatcherFromTyped[*slim_corev1.ServiceList](k8s.WatcherClient().CoreV1().Services(corev1.NamespaceAll)),
+		&slim_corev1.Service{},
+		0,
+		cache.ResourceEventHandlerFuncs{
```

Comment:
So, the modern way of writing these sorts of event queues is to not enqueue any kind of event. Rather, just enqueue the NamespacedName of the service (or whatever) that has changed. Then, the event loop retrieves the service (from the informer) and makes any changes.

This does rely on the ultimate logic being level driven, since you don't get a diff.

This has the advantage of coalescing multiple updates to the same object if its sitting in the queue.

Reply:
good point :100: , I remember that controller-runtime is following the above approach by default :100: